osc.lua: open select.lua by clicking buttons

guidocella opened the pull request:
8bf5548589 added mouse support to selectors, so open them from OSC buttons.

It's up for debate where middle or right click should be bound. By binding middle click `show_message(get_tracklist())` is no longer bound. By binding right click, it no longers cycles tracks backwards, though scrolling with the wheel still does.

Closes #5067, closes #6291, closes #10621, closes #11878.

Samillion commented:
I think right now choosing middle button is the right way, since it would be considered a sudden change.

It's very unfortunate, but many are not aware about `select.lua`, even when I tried to explain it, there is hesitation (fear of change), but once they see it in action, they never go back.

In time, can easily switch it to right click. Hell, lazy future scenario, make it a `user_opts`. :D

kasper93 commented:
Now that I think about it, mbtn_mid is not that discoverable, especially on mobile devices, like laptops. But making breaking changes for OSC, like changing mbtn_right might not be welcome by users, but maybe it is better thing to do?

Samillion commented:
It's a tough decision to make, any choice will have its downsides and caveats.

Getting it out there in a mid safe way is honestly a wise choice. Then wait for feedback and hear the dominating request, or more realistically, once it's actively used, the "choice" will become obvious.

Because, and I'm assuming here, this will be a solid replacement for `get_*list()` or `show-text *list` in the future.

If anything, this is a temporary confusion of a choice.

guidocella commented:
Rebinding `mbtn_mid` is still a breaking change, and changing defaults multiple times will annoy users more. Going to next/prev playlist entry/chapter is useful, but we could bind select to `mbtn_left` on the title where left click is less useful and move the current show-text function  to `mbtn_mid`, and on tracks where seeing what you're selecting beforehand is more useful lthan with playlist and chapter.

guidocella commented:
I can bind `MBTN_RIGHT` to closing the console or print circles in the playlist selector (this was requested before in https://github.com/mpv-player/mpv/pull/14087#issuecomment-2110889436) if desired.

Making everything configurable would be tricky since most clicks call internal functions rather than string commands and there are dozens of handlers.

kasper93 commented:
> Rebinding mbtn_mid is still a breaking change

I guess when changing UI code we cannot avoid changes. Else we will keep the same osc.lua forever.

guidocella commented:
> Lots of them already have associated script messages which can be invoked with `script-message-to`. In fact the replaced functionalities here are already covered by `osc-chapterlist`, `osc-playlist`, and `osc-tracklist` script messages. You can make these `open_selector` actions as script messages too.

I think we can add `osc-hide` and bind `script-binding select/select-playlist; script-message-to osc osc-hide` and so on. But in general I never understood why OSC has custom printing for playlists and chapters instead of just calling `show-text` (at least for the tracks it only prints tracks of that type). I see now that it even has custom printing after changing track in `set_track` (??) I'd rather just let the OSD show the text of `cycle sid` than to convert that to a script message as well. The track count can be added to that message.

> In fact, the only thing that would stop this from being "main" from the get-go, is [the font size scale to window issue](https://github.com/mpv-player/mpv/discussions/14903#discussioncomment-10794701). I simply added a [note](https://github.com/Samillion/ModernZ?tab=readme-ov-file#interactive-playlist) about it.

I can reopen https://github.com/mpv-player/mpv/pull/14088 if requested.

kasper93 commented:
> But in general I never understood why OSC has custom printing for playlists and chapters

Remove it, if you don't like it.

kasper93 reviewed (approved):
Works, as expected.

guidocella commented:
I thought we were going to make bindings configurable first?

kasper93 commented:
> I thought we were going to make bindings configurable first?

Ok, if you prefer to do this in this pr.

Samillion commented on player/lua/osc.lua:
```diff
@@ -1858,6 +1858,22 @@ layouts["topbar"] = function()
+    mp.command("script-binding select/select-" .. type)
+
+    if user_opts.visibility == "auto" then
+        osc_visible(false)
```
This is merely an opinion that might match the use case for you guys:

On my OSC fork, I'm hiding the osc `osc_visible(false)` without checking `if user_opts.visibility == auto`

**Reason:**
Like we talked before, the OSC is at the bottom and `select/console` also renders at the bottom. I instantly got reports that OSC interferes (ie: hovering over area)

Thought I'd let you know, just in case. Since it would make sense to hide osc on `open_selector` regardless of visibility mode, because you want the focus to be on `select` once it is triggered without interference until ESC/leave `select`.

CogentRedTester replied:
Isn't that what the `osc/margins` user_property is for? To allow console/select to draw text above the OSC?

Samillion replied:
Indeed it is, just thought I'd share it. I've opted for completely hiding osc while selector is on.

Like I said, merely an opinion that might match your use case.

guidocella replied:
OSC with visibility always does not interfere because the console is rendered above it so I don't understand your point. If users intentionally make the OSC always visible why should we override their preference?

Samillion replied:
👍 To each their own, just sharing an opinion

guidocella commented:
I changed the bindings to right click to be usable on laptops and because they are more useful than `show-text`.

I also added the fade animation when hiding the osc.

Left click on the title now opens the playlist selector and "show playlist position and length and full title" is moved to right click, because left clicking the title is not as useful as left clicking to go to adjacent playlist entries or chapters, and the feature is more discoverable if at least 1 binding is left click.

If we're willing to do a bigger change, we could bind left click to select tracks, because it is more useful to see what tracks you're selecting beforehand than with playlist and chapter navigation. uosc also opens its track menu on left click and doesn't bind cycling tracks at all. Any thoughs before making the bindings configurable?

guidocella commented:
Done, but you can still cycle tracks with the wheel.

kasper93 commented:
Works fine for me, I'm sure there are details that we can bikesheed, but there is not much interest in this. So I'm inclined to merge it to gather more feedback and we can work iteratively on this.

> I'd be very hesitant to merge this if there's no mouse-only way to cancel the current select.lua action. Partially addressed by https://github.com/mpv-player/mpv/pull/15145 which I'm fine with.

there is also a right click that cancels select.

@guidocella: what do you think about adding click action on `cache` element? Could be bind to stats.lua page 3.

guidocella commented:
Showing cache stats is cute but I'm not sure it would be useful it the cache seconds and the cached regions are already shown?

Samillion commented:
> I'd be very hesitant to merge this if there's no mouse-only way to cancel the current select.lua action. 

I was going to suggest to render controls similar to window controls with a  `[ X ] Close`, prepended, always top. 

I'm not sure how viable that would be, especially because you can predict that once this becomes more popular, one of requests would be opts for background-box/opacity for visibility in console/select and maybe even stats.

Poor Guido has been flooded with notes lately xD, though it's probably because it's a great addition

guidocella commented:
Well you can now close by clicking anywhere on the top line so an X would only be an indication.

You can already give the console and stats a background with `--osd-border-style=background-box`.

Samillion commented:
> You can already give the console and stats a background with `--osd-border-style=background-box`.

Yup, one of the first things I've added for osd since I started using select/console a lot.

```
osd-level=1
osd-font-size=32
osd-shadow-offset=4
osd-back-color=0.0/0.0/0.0/0.50
osd-border-style=background-box
osd-playing-msg="${!playlist-count==1:${playlist-pos-1}/${playlist-count}}"
```
